File: elixir/ch1/ex114.ex
# cd("c:/Users/Yohan/IdeaProjects/sicp/elixir/ch1")
# Load with import_file("filename")
# Run with Module.func(args)
defmodule Ex114 do
  @moduledoc false
  def biggest_denomination(1), do: 1
  def biggest_denomination(2), do: 5
  def biggest_denomination(3), do: 10
  def biggest_denomination(4), do: 25
  def biggest_denomination(5), do: 50

  def cc(0, _), do: 1
  def cc(amount, _) when amount < 0, do: 0
  def cc(_, 0), do: 0
  def cc(amount, kinds_of_coins), do:
    cc(amount, kinds_of_coins - 1) + cc(amount - biggest_denomination(kinds_of_coins), kinds_of_coins)

  def count_change(amount), do: cc(amount, 5)
end

# Ex114.count_change(100) == 292

File: python/ch1/ex114.py
import unittest


def biggest_denomination(n):
    if n == 1:
        return 1
    elif n == 2:
        return 5
    elif n == 3:
        return 10
    elif n == 4:
        return 25
    elif n == 5:
        return 50


def cc(amount, kinds_of_coins):
    if amount == 0:
        return 1
    elif amount < 0 or kinds_of_coins == 0:
        return 0
    else:
        return cc(amount, kinds_of_coins-1) + cc(amount-biggest_denomination(kinds_of_coins), kinds_of_coins)


def count_change(n):
    return cc(n, 5)


class MyTestCase(unittest.TestCase):
    def test1(self):
        self.assertEqual(count_change(0), 1)
        self.assertEqual(count_change(1), 1)
        self.assertEqual(count_change(2), 1)
        self.assertEqual(count_change(11), 4)
        self.assertEqual(count_change(100), 292)


if __name__ == '__main__':
    unittest.main()
